### ISSUE TITLE
Adding Query String Parameters added from Upstream Filter

### ISSUE THREAD

**Issue by DeanPoulin** [paultuckey/urlrewritefilter#187]
Hi,

I've added the urlrewritefilter v4.0.3 to my project and everything's working great. I have a simple `urlrewrite.xml` configuration for one path that looks like:

```xml
<urlrewrite use-query-string="true">
    <rule>
       <from>/api/(.+)</from>
       <to type="proxy">http://localhost:9074/api/$1</to>
    </rule>
</urlrewrite>
```

I have an upstream filter in the filter chain that's adding headers to the request successfully. It's doing this via subclassing HttpServletRequestWrapper and overriding the appropriate header methods.

```java
public class ApiSecurityHttpServletRequestWrapper extends HttpServletRequestWrapper {
    private static final String AUTH_HEADER_NAME = "X-LNAuth";

    private Map<String, String[]> securityParameters = null;
    private Map<String, String[]> allParameters = null;

    public ApiSecurityHttpServletRequestWrapper(ServletRequest request) {
        super((HttpServletRequest)request);
        securityParameters = generateSecurityParameters(request);
    }

    @Override
    public String getParameter(final String name) {
        String[] params = getParameterMap().get(name);

        if (ArrayUtils.isNotEmpty(params)) {
            return params[0];
        }

        return super.getParameter(name);
    }

    @Override
    public Map<String, String[]> getParameterMap() {
        if (allParameters == null) {
            allParameters = new TreeMap<String, String[]>();
            allParameters.putAll(super.getParameterMap());
            allParameters.putAll(securityParameters);
        }

        return allParameters;
    }

    @Override
    public Enumeration<String> getParameterNames() {
        return Collections.enumeration(getParameterMap().keySet());
    }

    @Override
    public String[] getParameterValues(final String name) {
        return getParameterMap().get(name);
    }

    private Map<String, String[]> generateSecurityParameters(ServletRequest request) {
        String key = EnvironmentConfig.getClientKey();
        String secret = EnvironmentConfig.getClientSecret();

        Map<String, String[]> parameters = new TreeMap<>();

        parameters.put("timestamp", new String[] {String.valueOf(System.currentTimeMillis())});
        parameters.put("key", new String[]{key});

        return parameters;
    }

    @Override
    public Enumeration<String> getHeaderNames() {
        ArrayList<String> headerNames = Collections.list(super.getHeaderNames());
        headerNames.add(AUTH_HEADER_NAME);

        return Collections.enumeration(headerNames);
    }
        
    @Override
    public String getHeader(String name) {
        if (AUTH_HEADER_NAME.equals(name)) {
            return createAuthHeader();
        }

        return super.getHeader(name);
    }

    @Override
    public Enumeration<String> getHeaders(String name) {
        if (AUTH_HEADER_NAME.equals(name)) {
            return Collections.enumeration(Arrays.asList(createAuthHeader()));
        }

        return super.getHeaders(name);
    }
        
    private String createAuthHeader() {
        UserInfo ui = UserInfo.fromAuthenticationData(DomainUsersInfo.getAuthenticationData());
        Gson gson = new Gson();
        String json = gson.toJson(ui);
        byte[] bytes = Base64.encodeBase64(json.getBytes());

        return new String(bytes);
    }
}
```

The headers are added, and the urlrewritefilter is passing the headers downstream, I'm happy.

I then got a new requirement. I need to have the upstream filter add query string parameters to the request that are also passed downstream.

I've overridden the correct methods in my subclassed HttpServletRequestWrapper class, but when I put breakpoints in those overridden methods, those methods are never called and my new parameters are not added to the request the urlrewritefilter sends downstream.

Do you have any suggestions on how to properly add query string parameters to a request the urlrewritefilter handles as a downstream filter that it will add to the query string?

**Comment by DeanPoulin** [comment]
@paultuckey - do you have a suggestion on this? Could use your insight. 😀

**Comment by paultuckey** [comment]
Hi, have you tried with `qsappend` on?

http://cdn.rawgit.com/paultuckey/urlrewritefilter/master/src/doc/manual/4.0/index.html#to

**Comment by DeanPoulin** [comment]
I will try that config and let you know thanks @paultuckey.

**Comment by DeanPoulin** [comment]
I have tried two configurations:

```xml
<urlrewrite use-query-string="true">
    <rule>
       <from>/api/(.+)</from>
       <to type="proxy" qsappend="true">http://localhost:9074/api/$1</to>
    </rule>
</urlrewrite>
```

With the above configuration, I'm seeing the query string being duplicated, so the original request into the application was:

http://local.myapp.com/api/init/app?appId=scr

The proxied request was:

http://localhost:9074/api/init/app?appId=scr&appid=scr

When I tried this configuration:

```xml
<urlrewrite>
    <rule>
       <from>/api/(.+)</from>
       <to type="proxy" qsappend="true">http://localhost:9074/api/$1</to>
    </rule>
</urlrewrite>
```

The proxied request was:

http://localhost:9074/api/init/app&appid=scr

It's still ignoring the additional parameters I added in the upstream filter as noted above. I must be doing something wrong.

I'm basically trying to have requests to my web app proxied through the UrlRewriteFilter to a downstream webserver with additional querystring parameters added to the request programatically since the downstream web server needs these additional parameters, but they are not supplied by the client.

**Comment by paultuckey** [comment]
I don't know sorry.  I can only suggest turning on full debug logging (or set breakpoints and use a debugger) for the filter and trying to figure out where it needs to be added and where.  Happy to accept patches!

**Comment by DeanPoulin** [comment]
sounds great. thanks for the suggestions. I will turn on debugging and step through and report the results. If I find something I will definitely report back here.

**Comment by DeanPoulin** [comment]
I did some debugging, and I found what's happening. The UrlRewriteFilter is calling to request.getQueryString() which I hadn't overridden (https://github.com/paultuckey/urlrewritefilter/blob/master/src/main/java/org/tuckey/web/filters/urlrewrite/UrlRewriter.java#L187).

I had improperly assumed that call would go through the get parameter and get parameters methods, so overriding the getQueryString method resolved the issue for me. I still left the other overrides so that the same behavior would occur if the downstream filters called getParameter on the new query string params I programmatically added.

Thanks so much for your suggestions, I'm going to close this since it's not an issue.
